kani-compiler/src/kani_middle/transform/check_aliasing/instrumentation.rs

```diff
@@ -0,0 +1,446 @@
+        let body = &mut self.body;
+        let local_ptr = self
+            .meta_stack
+            .entry(local)
+            .or_insert_with(|| body.new_local(ptr_ty, span, Mutability::Not));
+        let local_ptr = *local_ptr;
+        self.assign_pointer(local_ptr, local);
+        self.call(
+            Signature::new("KaniInitializeLocal", &[GenericArgKind::Type(ty)]),
+            vec![local_ptr],
+            self.unit,
+        )?;
+        Ok(())
+    }
+
+    /// Split at the minimum processed instruction,
+    /// allowing instrumentation of ghost code following
+    /// that source instruction.
+    pub fn process_instruction(&mut self) {
+        // If the instruction under processing is a terminator,
+        // special care is needed; it is impossible to instrument
+        // "after" a terminator with no target.
+        // Therefore we handle the terminators manually,
+        // inserting the terminator into the ghost code,
+        // then inserting a jump to that terminator.
+        // These will be called the "enter ghost,"
+        // "execute ghost", and "execute terminator"
+        // terminators
+        match self.min_processed {
+            SourceInstruction::Terminator { bb } => {
+                let original = self.body.blocks()[bb].terminator.clone();
+                let original_span = self.body.blocks()[bb].terminator.span.clone();
+                let span = self.span;
+                let kind = TerminatorKind::Goto { target: 0 };
+                let terminator = Terminator { kind, span };
+                let source = &mut self.min_processed.clone();
+                let enter_ghost_block = source.clone();
+                let body = &mut self.body;
+                body.replace_terminator(source, terminator.clone()); // replace terminator so you can instrument "after" it
+                body.insert_terminator(source, InsertPosition::After, terminator.clone());
+                let execute_terminator_block = source.clone();
+                body.insert_terminator(source, InsertPosition::After, terminator.clone());
+                let execute_ghost_block = source.clone();
+
+                // Instrument enter ghost:
+                let span = original_span;
+                let target = execute_ghost_block.bb();
+                let kind = TerminatorKind::Goto { target };
+                let terminator = Terminator { kind, span };
+                body.replace_terminator(&enter_ghost_block, terminator);
+                // Instrument execute ghost:
+                let target = execute_terminator_block.bb();
+                let kind = TerminatorKind::Goto { target };
+                let terminator = Terminator { kind, span };
+                body.replace_terminator(&execute_ghost_block, terminator);
+                // Instrument execute terminator
+                body.replace_terminator(&execute_terminator_block, original);
+
+                self.ghost_index = execute_ghost_block;
```

Comment:
It seems like what you are trying to do here is to insert a bunch of statements *before* the terminator in the basic block, so perhaps you should do it directly.